File: Lecture 05/Python Code/SRP/SRPViolated.py
class Product:
    def __init__(self, name, price):
        self.name = name
        self.price = price

class ShoppingCart:
    def __init__(self):
        self.products = []

    def add_product(self, product):
        self.products.append(product)

    def calculate_total(self):
        return sum(p.price for p in self.products)

    def print_invoice(self):
        print("Shopping Cart Invoice:")
        for product in self.products:
            print(f"{product.name} - Rs {product.price}")
        print(f"Total: Rs {self.calculate_total()}")

    def save_to_database(self):
        print("Saving shopping cart to database...")

if __name__ == "__main__":
    cart = ShoppingCart()
    cart.add_product(Product("Laptop", 50000))
    cart.add_product(Product("Mouse", 2000))

    cart.print_invoice()
    cart.save_to_database()

File: Lecture 05/JavaScript Code/SRP/SRP_violated.js
// Product class representing any item of any ECommerce.
class Product {
    constructor(name, price) {
        this.name = name;
        this.price = price;
    }
}

// Violating SRP: ShoppingCart is handling multiple responsibilities
class ShoppingCart {
    constructor() {
        this.products = [];
    }

    addProduct(product) {
        this.products.push(product);
    }

    getProducts() {
        return this.products;
    }

    // 1. Calculates total price in cart.
    calculateTotal() {
        let total = 0;
        for (const product of this.products) {
            total += product.price;
        }
        return total;
    }

    // 2. Violating SRP - Prints invoice (Should be in a separate class)
    printInvoice() {
        console.log("Shopping Cart Invoice:");
        for (const product of this.products) {
            console.log(`${product.name} - Rs ${product.price}`);
        }
        console.log(`Total: Rs ${this.calculateTotal()}`);
    }

    // 3. Violating SRP - Saves to DB (Should be in a separate class)
    saveToDatabase() {
        console.log("Saving shopping cart to database...");
    }
}

// Main execution
function main() {
    const cart = new ShoppingCart();

    cart.addProduct(new Product("Laptop", 50000));
    cart.addProduct(new Product("Mouse", 2000));

    cart.printInvoice();
    cart.saveToDatabase();
}

main();

